### PR TITLE
DAOS-8482 bio: set hotplug filter with pci bus id range

### PR DESCRIPTION
**Pull request by tanabarr**
Hotplug events are to be filtered so that an engine only responds to
events on PCI addresses with specific bus-ID values. This is to avoid
the situation where multiple per-engine SPDK hotplug pollers mistakenly
remove/plug the same device.

The hotplug filter configuration can be a hexadecimal bus ID range
e.g. 0-FF and will be set by the administrator in the server config file.

The range will be communicated to the engine by being written into the
SPDK JSON file that defines the bdev storage to be used by an engine.

The engine's bio module will then decode the range from JSON and
set the hotplug filter function (using the range as decider logic) to be
used to filter events detected by the SPDK hotplug poller.

Signed-off-by: Tom Nabarro <tom.nabarro@intel.com>

### CODE REVIEW

File: src/bio/bio_xstream.c [hunk 1]
```diff
@@ -73,6 +76,85 @@ struct bio_nvme_data {
...
+	int64_t		high;
+	int64_t		bus_id;
+
+	if (spdk_pci_addr_fmt(str, sizeof(str), addr) != 0) {
+		D_ERROR("unable to format pci address\n");
+		return false;
+	}
+
+	if ((!g_hotplug_bus_id_low) || (!g_hotplug_bus_id_high))
+		return false;
```
Comment (NiuYawei) [inline src/bio/bio_xstream.c:127]: I think filter function returning 'true' means the device should be included, 'false' means the device should be excluded, right? so I think filter function should return 'true' if there is no filter configured.

File: src/bio/bio_xstream.c [hunk 2]
```diff
@@ -73,6 +76,85 @@ struct bio_nvme_data {
...
+
+	low = hexstr_to_int(g_hotplug_bus_id_low);
```
Comment (NiuYawei) [inline src/bio/bio_xstream.c:129]: filter validation shouldn't be done in filter function, it should be done when setting filter.

File: src/bio/bio_xstream.c [hunk 3]
```diff
@@ -73,6 +76,85 @@ struct bio_nvme_data {
...
+	if (low < 0) {
+		D_ERROR("low bus id %s invalid\n", g_hotplug_bus_id_low);
+		return false;
+	}
+
+	high = hexstr_to_int(g_hotplug_bus_id_high);
```
Comment (NiuYawei) [inline src/bio/bio_xstream.c:135]: ditto.

File: src/bio/bio_xstream.c [hunk 4]
```diff
@@ -73,6 +76,85 @@ struct bio_nvme_data {
...
+	if (high < 0) {
+		D_ERROR("high bus id %s invalid\n", g_hotplug_bus_id_high);
+		return false;
+	}
+
+	bus_id = traddr_to_bus_id(str);
+	if (bus_id < 0) {
+		D_ERROR("input transport address %s invalid\n", str);
+		return false;
+	}
```
Comment (NiuYawei) [inline src/bio/bio_xstream.c:145]: I'm wondering if we should return 'true' to allow the device on error?

File: src/bio/bio_xstream.c [hunk 5]
```diff
@@ -73,6 +76,85 @@ struct bio_nvme_data {
...
+
+	D_INFO("hotplug: bus id %lX (filter %lX-%lX)\n", bus_id, low, high);
+	if ((bus_id < low) || (bus_id > high)) {
```
Comment (NiuYawei) [inline src/bio/bio_xstream.c:148]: Do we expect user to set filter for valid bus ID range or excluded bus ID range? I think setting valid (allowed) bus ID looks more straightforward?

**Comment by tanabarr** [comment]
With PR changes, reading busid range works from the following SPDK JSON config:
```
{
  "daos": {
    "config": [
      {
        "params": {
          "start": 0,
          "end": 131
        },
        "method": "hotplug_busid_range"
      }
    ]
  },
  "subsystems": [
    {
      "subsystem": "bdev",
      "config": [
        {
          "params": {
            "bdev_io_pool_size": 65536,
            "bdev_io_cache_size": 256
          },
          "method": "bdev_set_options"
        },
        {
          "params": {
            "retry_count": 4,
            "timeout_us": 0,
            "nvme_adminq_poll_period_us": 100000,
            "action_on_timeout": "none",
            "nvme_ioq_poll_period_us": 0
          },
          "method": "bdev_nvme_set_options"
        },
        {
          "params": {
            "enable": false,
            "period_us": 0
          },
          "method": "bdev_nvme_set_hotplug"
        },
        {
          "params": {
            "trtype": "PCIe",
            "name": "Nvme_wolf-167.wolf.hpdd.intel.com_0_1",
            "traddr": "0000:83:00.0"
          },
          "method": "bdev_nvme_attach_controller"
        }
      ]
    }
  ]
}
```

**Comment by tanabarr** [comment]
With the following server config:
```
name: daos_server           # sys group daos_server
access_points: ['wolf-167']
provider: ofi+sockets
port: 10001
nr_hugepages: 8192
control_log_mask: DEBUG
control_log_file: /tmp/daos_control_cpd.log
helper_log_file: /tmp/daos_admin_cpd.log
enable_hotplug: true
engines:
-
  targets: 4                # count of storage targets per each server
  pinned_numa_node: 0             # offset of the first core for service xstreams
  nr_xs_helpers: 0          # count of offload/helper xstreams per server
  fabric_iface: eth0        # map to OFI_INTERFACE=eth0
  fabric_iface_port: 31416  # map to OFI_PORT=31416
  log_mask: INFO,mgmt=DEBUG,bio=DEBUG
  log_file: /tmp/daos_engine_cpd.0.log
  env_vars:                 # influence DAOS IO Server behaviour by setting env variables
  - CRT_TIMEOUT=300
  - CRT_CREDIT_EP_CTX=0
  scm_mount: /mnt/daos0
  scm_class: ram
  scm_size: 16
  bdev_class: nvme
  bdev_list: ["0000:83:00.0"] #, "0000:84:00.0"]  # generate regular nvme.conf
  bdev_busid_range: "0x83-0x83"
```

Generated SPDK JSON config file with busid range specified and hot plug enabled:
```
[tanabarr@wolf-167 ~]$ cat /mnt/daos0/daos_nvme.conf
{
  "daos_data": {
    "config": [
      {
        "params": {
          "begin": 131,
          "end": 131
        },
        "method": "hotplug_busid_range"
      }
    ]
  },
  "subsystems": [
   {
      "subsystem": "bdev",
      "config": [
...
        {
          "params": {
            "enable": true,
            "period_us": 5000000
          },
          "method": "bdev_nvme_set_hotplug"
        },
...
``` 
Engine reports:
```
11/18-18:16:19.66 wolf-167 DAOS[185659/-1/0] bio  INFO src/bio/bio_config.c:656 bio_get_hotplug_busid_range() Hotplug bus-ID range read from config: 83-83
```
Which indicates correct range was set in spdk_set_hotplug_filter function parameter.

File: src/bio/bio_xstream.c [hunk 6]
```diff
@@ -108,6 +169,9 @@ bio_spdk_env_init(void)
 
 	opts.env_context = (char *)dpdk_cli_override_opts;
 
+	bio_get_hotplug_busid_range(nvme_glb.bd_nvme_conf);
+	spdk_nvme_pcie_set_hotplug_filter(hotplug_filter_fn);
```
Comment (NiuYawei) [inline src/bio/bio_xstream.c:173]: Minor: All these hotplug filter related code could be moved into bio_config.c, and export only single function like 'bio_set_hotplug_filter(nvme_conf)' in bio_internal.h.

**Review by NiuYawei — APPROVED** [review]

File: src/bio/bio_xstream.c (hunk 6 above)
Comment (tanabarr) [reply src/bio/bio_xstream.c]: this is a lot tidier thanks for the suggestion

File: utils/config/daos_server.yml [hunk 7]
```diff
@@ -337,6 +337,13 @@
 #    # PCIe addresses, and not the BDF format transport IDs of the backing NVMe SSDs
 #    # behind the VMD address. Also, 'disable_vmd' needs to be set to false.
 #    #bdev_list: ["0000:5d:05.5"]
+#
+#    # Filter hot-pluggable devices by PCI bus-ID by specifying a hexadecimal
+#    # range. Hotplug events relating to devices with PCI bus-IDs outside this range
+#    # will not be processed by this engine. Empty or unset range signifies allow all.
+#    bdev_busid_range: 0x80-0x8f
+#    #bdev_busid_range: 0x80-0x8F
+#    #bdev_busid_range: 128-143
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply utils/config/daos_server.yml:346]: to illustrate that number base will be inferred by prefix (or lack of), and that the hex values are caps-insensitive, I will remove the examples of the latter but keep the decimal example

**Review by NiuYawei — DISMISSED** [review]

**Review by NiuYawei — APPROVED** [review]

**Review by kjacque — APPROVED** [review]
Changes look okay to me.

**Review by mjmac — COMMENTED** [review]
What happens if the admin puts the same bus id range in multiple engine configs? Or enters the wrong bus id for an engine? The validation we're doing seems to protect against malformed values, but not nonsensical ones.

I still think that we should be able to do this automatically, rather than making the admin have to figure it out. If we know that engine 0 is on numa socket 0 and engine 1 is on numa socket 1, can't we figure out via hwloc which pci devices are eligible to be hot-plugged by a given engine?

**Comment by tanabarr** [comment]
> What happens if the admin puts the same bus id range in multiple engine configs? Or enters the wrong bus id for an engine? The validation we're doing seems to protect against malformed values, but not nonsensical ones.
> 
> I still think that we should be able to do this automatically, rather than making the admin have to figure it out. If we know that engine 0 is on numa socket 0 and engine 1 is on numa socket 1, can't we figure out via hwloc which pci devices are eligible to be hot-plugged by a given engine?

I am happy to add further validation in a follow-on PR. According to @NiuYawei we can't evaluate a "wrong" bus id for an engine for the same reason why we can't do this automatically ahead of time (as he has answered on the ticket). 

Can we please land this PR as it will unblock some of the hot plug tasks for @NiuYawei and @sydneyvanda and we can iterate on improvements as we develop the feature further? Thanks in advance.

**Comment by mjmac** [comment]
> I am happy to add further validation in a follow-on PR. According to @NiuYawei we can't evaluate a "wrong" bus id for an engine for the same reason why we can't do this automatically ahead of time (as he has answered on the ticket).
> 
As I noted in the ticket, I'm not convinced that there isn't a better way to define the filtering behavior, and I'm concerned that pushing on with the current design will lock us into an implementation that is harder to fix later. I would be happy to be shown that I'm wrong about this, though.

Relying on users to a) somehow know which values are correct b) remember to set them up seems like a recipe for disaster to me.

> Can we please land this PR as it will unblock some of the hot plug tasks for @NiuYawei and @sydneyvanda and we can iterate on improvements as we develop the feature further? Thanks in advance.

As far as I'm aware, master is still closed to non-2.0 landings, so I'm not sure what the push is here. It would be better to create a hotplug feature branch so that you can collaborate with others while the overall feature is in flux without impacting master stability.

**Comment by tanabarr** [comment]
replaced by https://github.com/daos-stack/daos/pull/7653 which integrates automatic range detection

**Pull request closed without merging**